SonarLint in Eclipse is slow due to failure in SonarTS

SonarTS will slowdown SonarLint analysis in Eclipse, because there is no typescript available.


```
No value provided by SonarLint for TypeScript location; property sonar.typescript.internal.typescriptLocation is missing
module.js:557
    throw err;
    ^

Error: Cannot find module 'typescript'
    at Function.Module._resolveFilename (module.js:555:15)
    at Function.Module._load (module.js:482:25)
    at Module.require (module.js:604:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/julien/Prog/workspace/.metadata/.plugins/org.eclipse.core.resources/.projects/sonar-branch-plugin/org.sonarlint.eclipse.core/.sonartmp/530654892781510723/sonarts-bundle/node_modules/tslint/lib/linter.js:20:10)
    at Module._compile (module.js:660:30)
    at Object.Module._extensions..js (module.js:671:10)
    at Module.load (module.js:573:32)
    at tryModuleLoad (module.js:513:12)
    at Function.Module._load (module.js:505:3)
Failed to start SonarTS Server
java.net.SocketTimeoutException: Accept timed out
        at java.net.PlainSocketImpl.socketAccept(Native Method) ~[na:1.8.0_144]
        at java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:409) ~[na:1.8.0_144]
        at java.net.ServerSocket.implAccept(ServerSocket.java:545) ~[na:1.8.0_144]
        at java.net.ServerSocket.accept(ServerSocket.java:513) ~[na:1.8.0_144]
        at org.sonar.plugin.typescript.ContextualServer.startSonarTSServer(ContextualServer.java:109) [sonar-typescript-plugin-1.5.0.2122.jar:na]
        at org.sonar.plugin.typescript.ContextualServer.start(ContextualServer.java:82) [sonar-typescript-plugin-1.5.0.2122.jar:na]
```

--- a/sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualSensor.java
+++ b/sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualSensor.java
@@ -52,6 +52,10 @@ public class ContextualSensor implements Sensor {
 
   @Override
   public void execute(SensorContext sensorContext) {
+    if (!contextualServer.isAlive()) {
+      LOG.warn("Skipped analysis as SonarTS Server is not running");
+      return;
+    }
     Iterable<InputFile> inputFiles = getInputFiles(sensorContext);
     LOG.info("Started SonarTS Analysis");
     inputFiles.forEach(inputFile -> {

--- a/sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java
+++ b/sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java
@@ -36,7 +36,6 @@ import org.sonar.api.config.Configuration;
 import org.sonar.api.utils.TempFolder;
 import org.sonar.api.utils.log.Logger;
 import org.sonar.api.utils.log.Loggers;
-import org.sonar.plugin.typescript.SensorContextUtils.AnalysisResponse;
 import org.sonar.plugin.typescript.executable.ExecutableBundle;
 import org.sonar.plugin.typescript.executable.ExecutableBundleFactory;
 import org.sonarsource.api.sonarlint.SonarLintSide;
@@ -94,11 +93,9 @@ public class ContextualServer implements Startable {
   }
 
   synchronized SensorContextUtils.AnalysisResponse analyze(SensorContextUtils.ContextualAnalysisRequest request) throws IOException {
-    if (!serverProcess.isAlive()) {
-      LOG.warn("Skipped analysis as SonarTS Server is not running");
-      return new AnalysisResponse();
+    if (!isAlive()) {
+      throw new IllegalStateException("ContextualServer is not started.");
     }
-
     final OutputStreamWriter writer = new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8);
     String requestJson = GSON.toJson(request);
     writer.append(requestJson);

--- a/sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java
+++ b/sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java
@@ -77,15 +77,20 @@ public class ContextualServer implements Startable {
 
   @Override
   public void start() {
-    externalProcessStreamConsumer.start();
-    LOG.info("Starting SonarTS Server");
+    Optional<String> typescriptLocation = configuration.get(TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY);
+    if (!typescriptLocation.isPresent()) {
+      LOG.warn("No value provided by SonarLint for TypeScript location; property " + TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY + " is missing");
+      LOG.warn("Skipping SonarTS Server start.");
+      return;
+    }
 
     if (isAlive()) {
       LOG.warn("Skipping SonarTS Server start, already running");
       return;
     }
 
-    startSonarTSServer();
+    LOG.info("Starting SonarTS Server");
+    startSonarTSServer(new File(typescriptLocation.get()));
   }
 
   synchronized SensorContextUtils.AnalysisResponse analyze(SensorContextUtils.ContextualAnalysisRequest request) throws IOException {
@@ -102,14 +107,15 @@ public class ContextualServer implements Startable {
     return GSON.fromJson(jsonReader, SensorContextUtils.AnalysisResponse.class);
   }
 
-  private void startSonarTSServer() {
+  private void startSonarTSServer(File typescriptLocation) {
     ExecutableBundle bundle = bundleFactory.createAndDeploy(tempFolder.newDir(), configuration);
     try {
       serverSocket = new ServerSocket(0);
       serverSocket.setSoTimeout(connectionTimeout);
       ProcessBuilder processBuilder = new ProcessBuilder(bundle.getSonarTSServerCommand(serverSocket.getLocalPort()).commandLineTokens());
-      setNodePath(processBuilder);
+      SensorContextUtils.setNodePath(typescriptLocation, processBuilder);
       serverProcess = processBuilder.start();
+      externalProcessStreamConsumer.start();
       externalProcessStreamConsumer.consumeStream(serverProcess.getErrorStream(), LOG::error);
       externalProcessStreamConsumer.consumeStream(serverProcess.getInputStream(), LOG::info);
       socket = serverSocket.accept();
@@ -119,15 +125,6 @@ public class ContextualServer implements Startable {
       if (isAlive()) {
         terminate();
       }
-    }
-  }
-
-  private void setNodePath(ProcessBuilder processBuilder) {
-    Optional<String> typescriptLocation = configuration.get(TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY);
-    if (typescriptLocation.isPresent()) {
-      SensorContextUtils.setNodePath(new File(typescriptLocation.get()), processBuilder);
-    } else {
-      LOG.warn("No value provided by SonarLint for TypeScript location; property " + TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY + " is missing");
     }
   }
 

--- a/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualSensorTest.java
+++ b/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualSensorTest.java
@@ -25,10 +25,12 @@ import org.junit.Test;
 import org.sonar.api.batch.fs.InputFile.Type;
 import org.sonar.api.batch.fs.internal.DefaultInputFile;
 import org.sonar.api.batch.fs.internal.TestInputFileBuilder;
+import org.sonar.api.batch.sensor.SensorContext;
 import org.sonar.api.batch.sensor.internal.DefaultSensorDescriptor;
 import org.sonar.api.batch.sensor.internal.SensorContextTester;
 import org.sonar.api.utils.internal.JUnitTempFolder;
 import org.sonar.api.utils.log.LogTester;
+import org.sonar.api.utils.log.LoggerLevel;
 import org.sonar.plugin.typescript.SensorContextUtils.AnalysisResponse;
 import org.sonar.plugin.typescript.SensorContextUtils.Issue;
 import org.sonar.plugin.typescript.SensorContextUtils.Position;
@@ -68,6 +70,7 @@ public class ContextualSensorTest {
     issue.name = inputFile.absolutePath();
     mockedResponse.issues[0] = issue;
 
+    when(contextualServer.isAlive()).thenReturn(true);
     when(contextualServer.analyze(any())).thenReturn(mockedResponse);
 
     ContextualSensor contextualSensor = new ContextualSensor(CHECK_FACTORY, contextualServer);
@@ -89,6 +92,7 @@ public class ContextualSensorTest {
     sensorContext.fileSystem().add(spiedInputFile);
 
     ContextualServer contextualServer = mock(ContextualServer.class);
+    when(contextualServer.isAlive()).thenReturn(true);
     verify(contextualServer, never()).analyze(any());
 
     ContextualSensor contextualSensor = new ContextualSensor(CHECK_FACTORY, contextualServer);
@@ -98,6 +102,17 @@ public class ContextualSensorTest {
     assertThat(logTester.logs()).contains("File with uri [https://example.org/foo/file.ts] can not be analyzed as it's not file scheme.");
   }
 
+  @Test
+  public void should_not_analyze_when_server_not_alive() throws Exception {
+    ContextualServer contextualServer = mock(ContextualServer.class);
+    when(contextualServer.isAlive()).thenReturn(false);
+    verify(contextualServer, never()).analyze(any());
+
+    ContextualSensor contextualSensor = new ContextualSensor(CHECK_FACTORY, contextualServer);
+    contextualSensor.execute(mock(SensorContext.class));
+
+    assertThat(logTester.logs(LoggerLevel.WARN)).contains("Skipped analysis as SonarTS Server is not running");
+  }
 
   @Test
   public void should_have_description() throws Exception {

--- a/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java
+++ b/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java
@@ -37,6 +37,7 @@ import org.sonar.plugin.typescript.SensorContextUtils.AnalysisResponse;
 import org.sonar.plugin.typescript.SensorContextUtils.ContextualAnalysisRequest;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
 import static org.sonar.plugin.typescript.TestUtils.BASE_DIR;
 import static org.sonar.plugin.typescript.TestUtils.TYPE_SCRIPT_RULES;
@@ -96,8 +97,7 @@ public class ContextualServerTest {
     contextualServer.start();
     assertThat(logTester.logs(LoggerLevel.ERROR)).contains("Failed to start SonarTS Server");
 
-    contextualServer.analyze(request);
-    assertThat(logTester.logs(LoggerLevel.WARN)).contains("Skipped analysis as SonarTS Server is not running");
+    assertThatThrownBy(() -> contextualServer.analyze(request)).isInstanceOf(IllegalStateException.class);
   }
 
   @Test

--- a/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java
+++ b/sonarts-sq-plugin/sonar-typescript-plugin/src/test/java/org/sonar/plugin/typescript/ContextualServerTest.java
@@ -81,8 +81,12 @@ public class ContextualServerTest {
     contextualServer.start();
 
     assertThat(logTester.logs(LoggerLevel.WARN))
-      .containsOnlyOnce("No value provided by SonarLint for TypeScript location; property sonar.typescript.internal.typescriptLocation is missing");
-    contextualServer.stop();
+      .containsOnlyOnce(
+        "No value provided by SonarLint for TypeScript location; property sonar.typescript.internal.typescriptLocation is missing",
+        "Skipping SonarTS Server start.")
+      .doesNotContain("Starting SonarTS Server");
+
+    assertThat(contextualServer.isAlive()).isFalse();
   }
 
   @Test

